### ISSUE TITLE
Deadlock / crash due to recursive singleton call 100% of first launches with FacebookAutoLogAppEventsEnabled 

### ISSUE THREAD

**Issue by davidhodge** [facebook/facebook-ios-sdk#1636]
## Checklist

- [x] I've updated to the latest released version of the SDK
- [x] I've searched for existing [GitHub issues](https://github.com/facebook/facebook-ios-sdk/issues)
- [x] I've looked for existing answers on [Stack Overflow](https://facebook.stackoverflow.com), the [Facebook Developer Community Forum](https://developers.facebook.com/community/) and the [Facebook Developers Group](https://www.facebook.com/groups/fbdevelopers)
- [x] I've read the [Code of Conduct](https://github.com/facebook/facebook-ios-sdk/blob/master/CODE_OF_CONDUCT.md)
- [x] This issue is not security related and can safely be disclosed publicly on GitHub

## Environment

Describe your dev environment here, giving as many details as possible. If you have them, make sure to include:

- Xcode Version: `12.4 12D4e`
- Installation Platform & Version: Cocoapods

App Configuration:
- Auto-logging of events is enabled in the info.plist with FacebookAutoLogAppEventsEnabled. Crash Will not occur without this due to FBSDKAppEvents:900
- I called activateApp, despite FacebookAutoLogAppEventsEnabled being enabled. I did this to work around another bug that was less serious, and I recognize this shouldn't *need* to be called. But it also shouldn't cause a crash.

## Goals

I want the FB SDK to allow install attribution, be allowed to explicitly call activeApp despite FacebookAutoLogAppEventsEnabled being enabled, and not experience a crash.

## Expected Results

I want the FB SDK to allow install attribution, be allowed to explicitly call activeApp despite FacebookAutoLogAppEventsEnabled being enabled, and not experience a crash.

## Actual Results

FB SDK Crashes due to FBSDKAppEvents singleton triggering a call to itself while still inside a dispatch_once.
<img width="666" alt="A80099AD-2EEC-4B20-B3B5-5B98FDA8C890-1933-00000140D1518879" src="https://user-images.githubusercontent.com/838040/106808075-c845c600-661e-11eb-96b3-a7c5e195177c.png">
<img width="234" alt="261A77E6-8C45-49FE-B405-C8BCDBFBCC70-1933-00000141FFC8D5AB" src="https://user-images.githubusercontent.com/838040/106808086-cbd94d00-661e-11eb-8e99-a3b481234ede.png">

## Steps to Reproduce

I have a code sample below. If you want to trigger it from scratch, the simplest way to reproduce this is to delete and re-install an app with auto-logging enabled. This will cause a crash on first launch. Re-install again to re-trigger.

Technically, FBSDKSettingsBitmask a.k.a @"com.facebook.sdk:FBSDKSettingsBitmask" needs to be different from the previous launch, but this happens every "first launch". This triggers logInternalEvent which starts the cascade of calls, which triggers the set of calls that ultimately result in the dispatch_once being hit again while it's still in progress.

Things tested:
FacebookAutoLogAppEventsEnabled:1, activateApp called directly: reliable crash on first launch
FacebookAutoLogAppEventsEnabled:1, activateApp not called directly: no crash
FacebookAutoLogAppEventsEnabled:0, activateApp called directly: no crash
FacebookAutoLogAppEventsEnabled:0, activateApp not called directly: no crash

## Code Samples & Details

I've attached a super simple code sample that crashes on every re-install on an iOS simulator. Key elements to note are the FacebookAutoLogAppEventsEnabled being 1 in the info.plist and the initialization of the SDK.

- I call  "[FBSDKAppEvents activateApp]" from my "application: willFinishLaunchingWithOptions:" and that triggers this every "first launch". I've tested this and it will also trigger from "didFinishLaunchingWithOptions"

[Nikola FB Sample.zip](https://github.com/facebook/facebook-ios-sdk/files/5921699/Nikola.FB.Sample.zip)

In conclusion, I do think it's a reasonable interpretation that I misused the SDK by having this flag set to 1 AND also  manually initializing the SDK. However, I am pretty firmly in the camp that such a seemingly trivial misuse shouldn't crash and certainly shouldn't not crash in one SDK and crash in a new one. (I've had this code for months without issue and now crashes with SDK v9)

Thanks for taking a look!

**Comment by joesus** [comment]
First off, thank you for the detailed writeup and code sample! 

> AND also manually initializing the SDK

It looks like the SDK is not being initialized anywhere. In v9 we removed auto-initialization which means the SDK will need to be either explicitly or implicitly initialized before use. Please see: https://github.com/facebook/facebook-ios-sdk/blob/master/CHANGELOG.md#removed for more details.

**Comment by davidhodge** [comment]
You're welcome! Regarding this point, I will investigate. I'll note that the SDK does still seem to pay some amount of attention to this flag (causing the bug).


> 
> It looks like the SDK is not being initialized anywhere. In v9 we removed auto-initialization which means the SDK will need to be either explicitly or implicitly initialized before use. Please see: https://github.com/facebook/facebook-ios-sdk/blob/master/CHANGELOG.md#removed for more details.

**Comment by davidhodge** [comment]
Oh I see what you mean here. So I'm not initializing the SDK explicitly, but I do have autologging, which is different.

**Comment by davidhodge** [comment]
Can confirm no crash after I do the right thing to initialize the SDK. Thank you

I suspect if you make this less crashy OR 100% crashy (not just on first launch), you will improve many lives! Cheers.

**Comment by joesus** [comment]
> Can confirm no crash after I do the right thing to initialize the SDK. Thank you
> 
> I suspect if you make this less crashy OR 100% crashy (not just on first launch), you will improve many lives! Cheers.

Awesome! A patch release in the near future should improve the crash-iness and log a helpful error message. :D
